broker-core/src/main/java/io/zeebe/broker/workflow/processor/flownode/IOMappingHelper.java

```diff
@@ -34,25 +36,42 @@
     final WorkflowInstanceRecord record = context.getValue();
     final ZeebeOutputBehavior outputBehavior = element.getOutputBehavior();
 
-    final DirectBuffer scopePayload =
-        context
-            .getElementInstanceState()
-            .getVariablesState()
-            .getVariablesAsDocument(record.getScopeInstanceKey());
     mergeTool.reset();
 
-    // TODO (saig0) #1613: if the activity has no output mappings then we don't need to propagate
-    // the variables
+    // (saig0): we need copy the buffer so that it is not overridden by the second variable request.
+    // This can be remove when #1852 is implemented.
+    final DirectBuffer scopePayload =
+        BufferUtil.cloneBuffer(
+            context
+                .getElementInstanceState()
+                .getVariablesState()
+                .getVariablesAsDocument(record.getScopeInstanceKey()));
+
+    final DirectBuffer instancePayload;
+
+    if (element instanceof ExecutableServiceTask && element.getOutputMappings().length > 0) {
+      // (saig0): if the activity has no output mappings then the payload is already merged. We
+      // can't use the local variables in this case for the record payload because it may contain
+      // variables from input mapping. This will go away together with the record payload.
+
+      instancePayload =
+          context
+              .getElementInstanceState()
+              .getVariablesState()
+              .getVariablesLocalAsDocument(context.getRecord().getKey());
+    } else {
+      // TODO (saig0) #1614: should use also the variables from the state when the message payload
+      // is set as local variables
+      instancePayload = record.getPayload();
```

Comment:
Correct me if I'm wrong, but I think in `JobCompletedProcessor` we overwrite the scope with the job payload if the task had no output mappings. So here if there was no output mappings we're effectively merging a payload on itself? I get that this is temporary though since hopefully everything will use variables and this can then be fixed, just want to make sure I understand the data flow.